p2p/book/book.go

```diff
@@ -0,0 +1,434 @@
+}
+
+func bucketize(raw Address) bucket {
+	if manet.IsPublicAddr(raw) || isDns(raw) {
+		return public
+	}
+	return private
+}
+
+type addressInfo struct {
+	// exported values will be persisted
+	ID        ID          `json:"id"`
+	Raw       jsonAddress `json:"raw"`
+	Class     class       `json:"class"`
+	Connected bool        `json:"connected"`
+
+	shareable bool // true if item is in shareable array
+	bucket    bucket
+	protected bool
+	failures  int
+	success   int
+}
+
+type jsonAddress struct {
+	Address
+}
+
+func (u *jsonAddress) UnmarshalJSON(data []byte) error {
+	// i didn't manage to find a way to use UnmarshalJSON method on the
+	// private multiaddr type
+	var s string
+	if err := json.Unmarshal(data, &s); err != nil {
+		return err
+	}
+	addr, err := ma.NewMultiaddr(s)
+	if err != nil {
+		return err
+	}
+	u.Address = addr
+	return nil
+}
+
+type Opt func(*Book)
+
+// WithLimit updates default book limit.
+func WithLimit(limit int) Opt {
+	return func(b *Book) {
+		b.limit = limit
+	}
+}
+
+// WithRand overwrites default seed for determinism in tests.
+func WithRand(seed int64) Opt {
+	return func(b *Book) {
+		b.rng = rand.New(rand.NewSource(seed))
+	}
+}
+
+func New(opts ...Opt) *Book {
+	b := &Book{
+		limit:     50000,
+		known:     map[ID]*addressInfo{},
+		queue:     list.New(),
+		rng:       rand.New(rand.NewSource(time.Now().Unix())),
+		shareable: []*addressInfo{},
+	}
+	for _, opt := range opts {
+		opt(b)
+	}
+	return b
+}
+
+type Book struct {
+	mu        sync.Mutex
+	limit     int
+	known     map[ID]*addressInfo
+	queue     *list.List
+	rng       *rand.Rand
+	shareable []*addressInfo
+}
+
+func (b *Book) Add(src, id ID, raw Address) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	addr := b.known[id]
+	bucket := bucketize(raw)
+	if src != SELF {
+		if source := b.known[src]; source == nil || source.bucket != bucket {
+			return
+		}
+	}
+	if addr == nil && len(b.known) == b.limit {
+		// future improvement is to find any stale and replace it with learned
+		return
+	} else if addr == nil {
+		addr = &addressInfo{
+			Raw:       jsonAddress{raw},
+			Class:     learned,
+			ID:        id,
+			bucket:    bucket,
+			shareable: true,
+		}
+		b.shareable = append(b.shareable, addr)
+		b.queue.PushBack(addr)
+		b.known[id] = addr
+	} else if addr.Raw.Address != raw {
+		addr.Raw.Address = raw
+		addr.bucket = bucket
+	}
+}
+
+type Event int
+
+const (
+	Protect Event = iota
+	Connected
+	Disconnected
+	Success
+	Fail
+)
+
+func (b *Book) Update(id ID, events ...Event) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	addr := b.known[id]
+	if addr == nil {
+		return
+	}
+	for _, event := range events {
+		switch event {
+		case Protect:
+			addr.protected = true
+			addr.Class = stable
+			if !addr.shareable {
+				addr.shareable = true
+				b.shareable = append(b.shareable, addr)
+			}
+		case Connected:
+			addr.Connected = true
+		case Disconnected:
+			addr.Connected = false
+		case Success, Fail:
+			b.queue.PushBack(addr)
```

Comment:
why put the addr back onto the queue when a crawl attempt had failed previously?

Reply:
it is dropped from a queue only when address is deleted from a book